cmd/containerd/command/service_windows.go

```diff
@@ -142,13 +171,39 @@ func registerService() error {
 		DisplayName:  "Containerd",
 		Description:  "Container runtime",
 	}
-
+	containerdProcessHandle := windows.CurrentProcess()
+	priority := uint32(0)
 	// Configure the service to launch with the arguments that were just passed.
 	args := []string{"--run-service"}
 	for _, a := range os.Args[1:] {
-		if a != "--register-service" && a != "--unregister-service" {
+		if a != "--register-service" && a != "--unregister-service" && a != "--priorityclass" {
 			args = append(args, a)
 		}
+		if strings.Contains(a, "--priorityclass") {
```

Comment:
`registerService` is only called to actually register the service with the Windows service manager, usually at install time. It's not called at containerd run-time.

So I think instead of this code, we can just add code under the `if runServiceFlag {` branch of `registerUnregisterService` below. We shouldn't need to parse the flag ourselves either, we can just get the argument value and use it directly as we do for other flags.

Reply:
Yup, I just realized that today. I've added the flag to `applyPlatformSpecificFlags` method and modified the signature of the method to return an error. Let me know if you're fine with it.

I also realized that this split of registering a windows service and starting the windows service is causing a problem while setting priority as we can change priority associated with running processes, so whatever changes we make in the code are having no effect as containerd process is started via `Start-Service` command manually after service has been registered. 

I am trying to understand history here - What is the reason for registering the containerd windows service without starting it?

Reply:
@kevpar - Can you please answer this ^?

> Here's some sample code that shows what it would look like to create a job object, set a priority class on it, and add the current process to the job:

Thanks for sharing the snippet. I agree that having a job object to manage containerd+shims is ideal going forward. Is there an interface that all the shims need to confirm to? If yes, we can have a method there to get the containerd process priority or perhaps something like `GetContainerdJobObject`?

Reply:
Windows separates out the operations of registering/unregistering a service, from actually starting the service. Containerd has the code for all of these in `service_windows.go`, but there are different paths through the code based on what operation is being performed. Once the service is registered, it is started separately through something like `net start` or `Start-Service`.

Basically, the Windows service code in containerd has 3 different cases that it has to handle:
- Registering containerd as a service (`--register-service` is passed). This would be done when containerd is installed on the system.
- Unregistering containerd as a service (`--unregister-service` is passed). This would be done when containerd is uninstalled on the system.
- Actually running as a service (`--run-service` is passed). This would be done when containerd.exe is invoked from the Windows service manager.

When invoked for registering or unregistering, the containerd instance is very short-lived. It just carries out the task of talking to the Windows service manager to make the required changes to the service configuration, and then exits. We don't need to set any process priority in these cases, since it's a long-lived containerd process, and is only invoked in special cases.

When invoked for running as a service, we _should_ set process priority, though. I think the best place to do this is inside the `if runServiceFlag` block [here](https://github.com/containerd/containerd/blob/c4526c98700147c12268b1f70c6b50cfd328c303/cmd/containerd/command/service_windows.go#L266).

> @kevpar - Can you please answer this ^?
> 
> > Here's some sample code that shows what it would look like to create a job object, set a priority class on it, and add the current process to the job:
> 
> Thanks for sharing the snippet. I agree that having a job object to manage containerd+shims is ideal going forward. Is there an interface that all the shims need to confirm to? If yes, we can have a method there to get the containerd process priority or perhaps something like `GetContainerdJobObject`?

There are some command line parameters that are expected to be recognized by shims, but we shouldn't need to worry about that. If we set containerd to be in a job object, then that membership will also be inherited by the shim child processes, so they will automatically inherit the process priority set on the job object.